fuzzing/nxt_basic_fuzz.c

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) NGINX, Inc.
+ */
+
+#include <nxt_main.h>
+
+
+#define KMININPUTLENGTH 2
+#define KMAXINPUTLENGTH 128
+
+
+extern int LLVMFuzzerInitialize(int *argc, char ***argv);
+extern int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size);
+
+void nxt_base64_fuzz(const u_char *data, size_t size);
+void nxt_term_fuzz(const u_char *data, size_t size);
+void nxt_time_fuzz(const u_char *data, size_t size);
+void nxt_utf8_fuzz(const u_char *data, size_t size);
+
+
+extern char  **environ;
+
+
+int
+LLVMFuzzerInitialize(int *argc, char ***argv)
+{
+    if (nxt_lib_start("fuzzing", NULL, &environ) != NXT_OK) {
+        return NXT_ERROR;
+    }
+
+    return 0;
+}
+
+
+int
+LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
+{
+    if (size < KMININPUTLENGTH || size > KMAXINPUTLENGTH) {
+        return 0;
+    }
+
+    nxt_base64_fuzz(data, size);
+    nxt_term_fuzz(data, size);
+    nxt_time_fuzz(data, size);
+    nxt_utf8_fuzz(data, size);
+
+    return 0;
+}
+
+
+void
+nxt_base64_fuzz(const u_char *data, size_t size)
+{
+    u_char  *input;
+    u_char  buf[256];
+
+    /*
+     * Adding unnecessary NUL bytes to resolve buffer-overflow read and
+     * use-of-uninitialized-value in nxt_base64_decode nxt_string.c:833:21.
+     */
+
+    input = nxt_zalloc(size + 1);
+    nxt_memcpy(input, data, size);
```

Comment:
There is no other way around. I tried to fix it by myself, but I am unable to pass the tests.

Reply:
I'd like to understand what exactly the problem is here...

Reply:
Alright, 

There is a bug in the `nxt_base64_decode` function at [nxt_string.c#L833](https://github.com/nginx/unit/blob/master/src/nxt_string.c#L833).

I already reported this to `"F5-SIRT"`. They said it's not a security vulnerability. Also, I'm aware that it's not a security bug. Usually, one or two bytes of over-read can't do much.

#### Reproduce:
Download [crash-fuzz_basic.txt](https://github.com/user-attachments/files/15612584/crash-fuzz_basic.txt).

```bash
git clone --depth=1 -b fuzzing-update https://github.com/pkillarjun/unit.git
cd unit
bash fuzzing/build-fuzz.sh
./build/fuzz_basic crash-fuzz_basic.txt # Can't upload .raw.
```

ASAN
```txt
=================================================================
==268367==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x502000000035 at pc 0x00000059f198 bp 0x7fffbfd3eb00 sp 0x7fffbfd3eaf8
READ of size 1 at 0x502000000035 thread T0
    #0 0x59f197 in nxt_base64_decode /home/oblivion/Desktop/unit/src/nxt_string.c:832:24
    #1 0x5454ed in nxt_base64_fuzz /home/oblivion/Desktop/unit/fuzzing/nxt_basic_fuzz.c:65:5
    #2 0x5454ed in LLVMFuzzerTestOneInput /home/oblivion/Desktop/unit/fuzzing/nxt_basic_fuzz.c:42:5
    #3 0x451a44 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/oblivion/Desktop/unit/build/fuzz_basic+0x451a44) (BuildId: 34d7d355b8590bdd92fee196ddf89ef7e3735913)
    #4 0x43a89a in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/home/oblivion/Desktop/unit/build/fuzz_basic+0x43a89a) (BuildId: 34d7d355b8590bdd92fee196ddf89ef7e3735913)
    #5 0x4401e0 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/oblivion/Desktop/unit/build/fuzz_basic+0x4401e0) (BuildId: 34d7d355b8590bdd92fee196ddf89ef7e3735913)
    #6 0x46bf56 in main (/home/oblivion/Desktop/unit/build/fuzz_basic+0x46bf56) (BuildId: 34d7d355b8590bdd92fee196ddf89ef7e3735913)
    #7 0x7faaf8810087 in __libc_start_call_main (/lib64/libc.so.6+0x2a087) (BuildId: 4a92fcedbba6d6d2629ce066a2970017faa9995e)
    #8 0x7faaf881014a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x2a14a) (BuildId: 4a92fcedbba6d6d2629ce066a2970017faa9995e)
    #9 0x434eb4 in _start (/home/oblivion/Desktop/unit/build/fuzz_basic+0x434eb4) (BuildId: 34d7d355b8590bdd92fee196ddf89ef7e3735913)
```

### Thought

After some debugging, I found out that this happened due to `src[1]` because there was not enough buffer available. So now the `NUL byte` becomes the wanted buffer, and `nxt_base64_decode` does not crash.

If i try to fix this over-read using `if (pad > 0 && end - src > 1)`, `nxt_base64_test` test dose not pass.

Reply:
> Alright,
> 
> There is a bug in the `nxt_base64_decode` function at [nxt_string.c#L833](https://github.com/nginx/unit/blob/master/src/nxt_string.c#L833).

Line 832 no? (according to the asan report)

> I already reported this to `"F5-SIRT"`. They said it's not a security vulnerability. Also, I'm aware that it's not a security bug. Usually, one or two bytes of over-read can't do much.

Yes, I saw that, but couldn't see an obvious issue with the code. I also ran our `./src/test/nxt_base64_test.c` under valgrind and libasan and didn't see any issue.

When trying to debug it via `./fuzzing/nxt_basic_fuzz.c` from what I could tell it was failing when passing in an empty buffer but with a size of 2.

Is there any way to actually see what the inputs are when it fails? i.e what is it that's being passed into `nxt_base64_decode()` for `src`, `dst` and `size`?

(Other than putting in a printf(), which is where I'm seeing the empty buffer...)

> #### Reproduce:
> 
> Download [crash-fuzz_basic.txt](https://github.com/user-attachments/files/15612584/crash-fuzz_basic.txt).
> 
> ```shell
> git clone --depth=1 -b fuzzing-update https://github.com/pkillarjun/unit.git
> cd unit
> bash fuzzing/build-fuzz.sh
> ./build/fuzz_basic crash-fuzz_basic.txt # Can't upload .raw.
> ```

```
$ hexdump -C ~/Downloads/crash-fuzz_basic.txt
00000000  2c e5 00 00 00                                    |,....|
00000005
```

Which seems to be a comma followed by the non-printable extended ASCII character 229, then three nuls.

> ASAN
> 
> ```
> =================================================================
> ==268367==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x502000000035 at pc 0x00000059f198 bp 0x7fffbfd3eb00 sp 0x7fffbfd3eaf8
> READ of size 1 at 0x502000000035 thread T0
>     #0 0x59f197 in nxt_base64_decode /home/oblivion/Desktop/unit/src/nxt_string.c:832:24
> ```

```c
        v1 = decode[src[0]];                                                   
```

> 
> ### Thought
> 
> After some debugging, I found out that this happened due to `src[1]` because there was not enough buffer available. So now the `NUL byte` becomes the wanted buffer, and `nxt_base64_decode` does not crash.
> 
> If i try to fix this over-read using `if (pad > 0 && end - src > 1)`, `nxt_base64_test` test dose not pass.
> 
> For Example:
> 
>     * `openssl base64 -d <<< ""z9+Npe=="" | xxd -p`
>       Output : `cfdf8da5`.
> 
>     * What is should be from `tests[i].dec.start`
>       Output : `cfdf 8da5`
> 
>     * After Patch.
>       Output : `cfdf 8d`

So if I use `./src/test/nxt_base64_test.c` with a `src` of `,\E5\00\00\00`, what `size/length` was it using, 5? and run that under valgrind I should see the issue?

Reply:
OK, a buffer with 13 chars fails, I suspect 17, 21, 25 ... will also fail... increments of 4 seems an issue... there is some mod 4 stuff going on in there...

Reply:
With a sbuf of a given size and the second src@ & end@ are just before we do decode[src[0]] & decode[src[1]]

```
sbuf[17]
src@0x402220, end@0x402231
pad [3] length [17]
src@0x402230, end@0x402230

sbuf[18]
src@0x402220, end@0x402232
pad [2] length [18]
src@0x402230, end@0x402230

sbuf[19]
src@0x402220, end@0x402233
pad [1] length [19]
src@0x402230, end@0x402230

sbuf[20]
src@0x402220, end@0x402234
pad [0] length [20]
src@0x402234, end@0x402234
```

When pad ==3 (which it will be for any length >= 4 and an increment of 4) the src address is only one away from the end address.

Seeing if we need to special case the pad == 3 case inside the if (pad > 0)...

Reply:
Here's a question.

Is it even possible for valid base64 to have 3 padding characters?

Reply:
I did have a patch handling the pad == 3 case, but if I'd looked harder further up the function, I would have noticed that we indeed do treat a padding of 3 as invalid...

```c
    if (dst == NULL) {                                                          
        if (pad > 2) {                                                          
            return NXT_ERROR;                                                   
        } 
```
So, yeah as I suspected a padding of 3 would mean invalid base64.

`nxt_base64_decode()` is only used by the OpenSSL code. The configuration validation code calls it once with dst as `NULL` for checking the base64 stored in the config.

There is also a single call to `nxt_base64_decode(`) in `nxt_tls_ticket_keys()` (which depends on NXT_HAVE_OPENSSL_TLSEXT & SSL_CTRL_SET_TLSEXT_TICKET_KEY_CB) where it retrieves some base64 from the config.

But that looks to be checked by the afore mentioned configuration check in `nxt_conf_vldt_ticket_key_element()`

```c
    ret = nxt_base64_decode(NULL, key.start, key.length);                       
    if (ret == NXT_ERROR) {                                                     
        return nxt_conf_vldt_error(vldt, "Invalid Base64 format for the ticket "
                                   "key \"%V\".", &key);                        
    }
```

So there is no issue for us.

The correct way to use this API is to call `nxt_base64_decode()` with `dst = NULL` first to check the validity of the base64.

Reply:
> The correct way to use this API is to call nxt_base64_decode() with dst = NULL first to check the validity of the base64.

Ya, thanks for the help.